File: src/main/java/com/perforce/performance/metrics/PerformanceMetrics.java
package com.perforce.performance.metrics;

/**
 * PerformanceMetrics - Data model holding all collected metrics for a single page navigation.
 */
public class PerformanceMetrics {

    private String pageName;
    private String pageUrl;
    private long timestamp;

    // ── Navigation Timing (milliseconds) ──────────────────────────────────────
    private double dnsLookupTime;
    private double tcpConnectionTime;
    private double timeToFirstByte;          // TTFB
    private double domContentLoadedTime;     // DCL
    private double pageLoadTime;             // Full load
    private double domInteractiveTime;
    private double redirectTime;
    private double requestTime;
    private double responseTime;

    // ── Core Web Vitals ────────────────────────────────────────────────────────
    private double largestContentfulPaint;   // LCP  (target: ≤ 2500ms)
    private double firstContentfulPaint;     // FCP  (target: ≤ 1800ms)
    private double cumulativeLayoutShift;    // CLS  (target: ≤ 0.1)
    private double interactionToNextPaint;   // INP  (target: ≤ 200ms)
    private double firstInputDelay;          // FID  (target: ≤ 100ms)
    private double totalBlockingTime;        // TBT  (target: ≤ 200ms)

    // ── Resource Metrics ──────────────────────────────────────────────────────
    private int totalResources;
    private long transferSize;               // bytes
    private double speedIndex;

    // ── Time to Interactive ───────────────────────────────────────────────────
    private double timeToInteractive;        // TTI  (target: ≤ 3800ms)

    // ── AG-Grid Spinner Time ──────────────────────────────────────────────────
    private double agGridLoadTime;           // ms spinner was visible (0 = no grid on page)

    // ── Benchmark Status ──────────────────────────────────────────────────────
    private String lcpStatus;
    private String fcpStatus;
    private String clsStatus;
    private String ttfbStatus;
    private String pageLoadStatus;
    private String ttiStatus;
    private String overallStatus;

    // ── Screenshot Path ───────────────────────────────────────────────────────
    private String screenshotPath;

    public PerformanceMetrics() {
        this.timestamp = System.currentTimeMillis();
    }

    // ─── Benchmark Evaluation ─────────────────────────────────────────────────

    public void evaluateBenchmarks() {
        this.lcpStatus       = evaluateLCP(largestContentfulPaint);
        this.fcpStatus       = evaluateFCP(firstContentfulPaint);
        this.clsStatus       = evaluateCLS(cumulativeLayoutShift);
        this.ttfbStatus      = evaluateTTFB(timeToFirstByte);
        this.pageLoadStatus  = evaluatePageLoad(pageLoadTime);
        this.ttiStatus       = evaluateTTI(timeToInteractive);

        // Overall = worst of all
        boolean allGood = "GOOD".equals(lcpStatus) && "GOOD".equals(fcpStatus)
                && "GOOD".equals(clsStatus) && "GOOD".equals(ttfbStatus)
                && "GOOD".equals(pageLoadStatus) && "GOOD".equals(ttiStatus);
        boolean anyPoor = "POOR".equals(lcpStatus) || "POOR".equals(fcpStatus)
                || "POOR".equals(clsStatus) || "POOR".equals(ttfbStatus)
                || "POOR".equals(pageLoadStatus) || "POOR".equals(ttiStatus);
        this.overallStatus = allGood ? "GOOD" : anyPoor ? "POOR" : "NEEDS IMPROVEMENT";
    }

    // Google / Industry Benchmarks
    private String evaluateLCP(double lcp) {
        if (lcp <= 0) return "N/A";
        return lcp <= 2500 ? "GOOD" : lcp <= 4000 ? "NEEDS IMPROVEMENT" : "POOR";
    }

    private String evaluateFCP(double fcp) {
        if (fcp <= 0) return "N/A";
        return fcp <= 1800 ? "GOOD" : fcp <= 3000 ? "NEEDS IMPROVEMENT" : "POOR";
    }

    private String evaluateCLS(double cls) {
        if (cls < 0) return "N/A";
        return cls <= 0.1 ? "GOOD" : cls <= 0.25 ? "NEEDS IMPROVEMENT" : "POOR";
    }

    private String evaluateTTFB(double ttfb) {
        if (ttfb <= 0) return "N/A";
        return ttfb <= 800 ? "GOOD" : ttfb <= 1800 ? "NEEDS IMPROVEMENT" : "POOR";
    }

    private String evaluatePageLoad(double load) {
        if (load <= 0) return "N/A";
        return load <= 3000 ? "GOOD" : load <= 6000 ? "NEEDS IMPROVEMENT" : "POOR";
    }

    private String evaluateTTI(double tti) {
        if (tti <= 0) return "N/A";
        return tti <= 3800 ? "GOOD" : tti <= 7300 ? "NEEDS IMPROVEMENT" : "POOR";
    }

    // ─── Getters / Setters ────────────────────────────────────────────────────

    public String getPageName()                    { return pageName; }
    public void setPageName(String pageName)       { this.pageName = pageName; }

    public String getPageUrl()                     { return pageUrl; }
    public void setPageUrl(String pageUrl)         { this.pageUrl = pageUrl; }

    public long getTimestamp()                     { return timestamp; }
    public void setTimestamp(long timestamp)       { this.timestamp = timestamp; }

    public double getDnsLookupTime()               { return dnsLookupTime; }
    public void setDnsLookupTime(double v)         { this.dnsLookupTime = v; }

    public double getTcpConnectionTime()           { return tcpConnectionTime; }
    public void setTcpConnectionTime(double v)     { this.tcpConnectionTime = v; }

    public double getTimeToFirstByte()             { return timeToFirstByte; }
    public void setTimeToFirstByte(double v)       { this.timeToFirstByte = v; }

    public double getDomContentLoadedTime()        { return domContentLoadedTime; }
    public void setDomContentLoadedTime(double v)  { this.domContentLoadedTime = v; }

    public double getPageLoadTime()                { return pageLoadTime; }
    public void setPageLoadTime(double v)          { this.pageLoadTime = v; }

    public double getDomInteractiveTime()          { return domInteractiveTime; }
    public void setDomInteractiveTime(double v)    { this.domInteractiveTime = v; }

    public double getRedirectTime()                { return redirectTime; }
    public void setRedirectTime(double v)          { this.redirectTime = v; }

    public double getRequestTime()                 { return requestTime; }
    public void setRequestTime(double v)           { this.requestTime = v; }

    public double getResponseTime()                { return responseTime; }
    public void setResponseTime(double v)          { this.responseTime = v; }

    public double getLargestContentfulPaint()      { return largestContentfulPaint; }
    public void setLargestContentfulPaint(double v){ this.largestContentfulPaint = v; }

    public double getFirstContentfulPaint()        { return firstContentfulPaint; }
    public void setFirstContentfulPaint(double v)  { this.firstContentfulPaint = v; }

    public double getCumulativeLayoutShift()       { return cumulativeLayoutShift; }
    public void setCumulativeLayoutShift(double v) { this.cumulativeLayoutShift = v; }

    public double getInteractionToNextPaint()      { return interactionToNextPaint; }
    public void setInteractionToNextPaint(double v){ this.interactionToNextPaint = v; }

    public double getFirstInputDelay()             { return firstInputDelay; }
    public void setFirstInputDelay(double v)       { this.firstInputDelay = v; }

    public double getTotalBlockingTime()           { return totalBlockingTime; }
    public void setTotalBlockingTime(double v)     { this.totalBlockingTime = v; }

    public int getTotalResources()                 { return totalResources; }
    public void setTotalResources(int v)           { this.totalResources = v; }

    public long getTransferSize()                  { return transferSize; }
    public void setTransferSize(long v)            { this.transferSize = v; }

    public double getSpeedIndex()                  { return speedIndex; }
    public void setSpeedIndex(double v)            { this.speedIndex = v; }

    public double getTimeToInteractive()             { return timeToInteractive; }
    public void setTimeToInteractive(double v)      { this.timeToInteractive = v; }

    public double getAgGridLoadTime()               { return agGridLoadTime; }
    public void setAgGridLoadTime(double v)         { this.agGridLoadTime = v; }

    public String getLcpStatus()                   { return lcpStatus; }
    public String getFcpStatus()                   { return fcpStatus; }
    public String getClsStatus()                   { return clsStatus; }
    public String getTtfbStatus()                  { return ttfbStatus; }
    public String getPageLoadStatus()              { return pageLoadStatus; }
    public String getTtiStatus()                   { return ttiStatus; }
    public String getOverallStatus()               { return overallStatus; }

    public String getScreenshotPath()             { return screenshotPath; }
    public void setScreenshotPath(String value)   { this.screenshotPath = value; }

    @Override
    public String toString() {
        return String.format(
            "[%s] Load=%.0fms | FCP=%.0fms | LCP=%.0fms | TTI=%.0fms | CLS=%.3f | TTFB=%.0fms | AG-Grid=%.0fms | Overall=%s",
            pageName, pageLoadTime, firstContentfulPaint, largestContentfulPaint,
            timeToInteractive, cumulativeLayoutShift, timeToFirstByte, agGridLoadTime, overallStatus
        );
    }
}

File: src/metrics/PerformanceMetrics.ts
import { ApiRequestMetric } from './ApiRequestMetric';
import { LighthouseMetrics } from './LighthouseMetrics';

export type BenchmarkStatus = 'GOOD' | 'NEEDS IMPROVEMENT' | 'POOR' | 'N/A';

/**
 * PerformanceMetrics — data model holding all collected metrics for a single page navigation.
 * Plain object (no class methods on instances); use module-level helpers.
 */
export interface PerformanceMetrics {
  pageName: string;
  pageUrl: string;
  timestamp: number;

  // Navigation Timing (ms)
  dnsLookupTime: number;
  tcpConnectionTime: number;
  timeToFirstByte: number;
  domContentLoadedTime: number;
  pageLoadTime: number;
  domInteractiveTime: number;
  redirectTime: number;
  requestTime: number;
  responseTime: number;

  // Core Web Vitals
  largestContentfulPaint: number;
  firstContentfulPaint: number;
  cumulativeLayoutShift: number;
  interactionToNextPaint: number;
  firstInputDelay: number;
  totalBlockingTime: number;

  // Resource
  totalResources: number;
  transferSize: number;  // bytes
  speedIndex: number;

  // TTI
  timeToInteractive: number;

  // AG-Grid spinner duration (ms, 0 = no grid on this page)
  agGridLoadTime: number;

  /**
   * Wall-clock time from the moment the navigation action fires (click or goto)
   * to when the page readiness handler confirms the page is fully interactive.
   * Measured by NavigationRouter.navigateAndPrepare() using Date.now().
   * Unlike pageLoadTime (which uses browser navigationStart), this captures
   * the full end-to-end latency including click-handler processing time.
   */
  actionToLoadMs: number;

  // Slow API requests (>= 300 ms)
  slowApiRequests: ApiRequestMetric[];

  // Benchmark statuses
  lcpStatus: BenchmarkStatus;
  fcpStatus: BenchmarkStatus;
  clsStatus: BenchmarkStatus;
  ttfbStatus: BenchmarkStatus;
  pageLoadStatus: BenchmarkStatus;
  ttiStatus: BenchmarkStatus;
  overallStatus: BenchmarkStatus;

  screenshotPath?: string;

  // Optional Lighthouse audit results (populated when running via PageLoadAnalyzerAgent)
  lighthouseData?: LighthouseMetrics;
}

// ── Factory ────────────────────────────────────────────────────────────────────

export function createEmptyMetrics(): PerformanceMetrics {
  return {
    pageName: '',
    pageUrl: '',
    timestamp: Date.now(),
    dnsLookupTime: 0,
    tcpConnectionTime: 0,
    timeToFirstByte: 0,
    domContentLoadedTime: 0,
    pageLoadTime: 0,
    domInteractiveTime: 0,
    redirectTime: 0,
    requestTime: 0,
    responseTime: 0,
    largestContentfulPaint: 0,
    firstContentfulPaint: 0,
    cumulativeLayoutShift: 0,
    interactionToNextPaint: 0,
    firstInputDelay: 0,
    totalBlockingTime: 0,
    totalResources: 0,
    transferSize: 0,
    speedIndex: 0,
    timeToInteractive: 0,
    agGridLoadTime: 0,
    actionToLoadMs: 0,
    slowApiRequests: [],
    lcpStatus: 'N/A',
    fcpStatus: 'N/A',
    clsStatus: 'N/A',
    ttfbStatus: 'N/A',
    pageLoadStatus: 'N/A',
    ttiStatus: 'N/A',
    overallStatus: 'N/A',
  };
}

// ── Benchmark evaluation ───────────────────────────────────────────────────────

export function evaluateBenchmarks(m: PerformanceMetrics): void {
  m.lcpStatus      = evalLCP(m.largestContentfulPaint);
  m.fcpStatus      = evalFCP(m.firstContentfulPaint);
  m.clsStatus      = evalCLS(m.cumulativeLayoutShift);
  m.ttfbStatus     = evalTTFB(m.timeToFirstByte);
  m.pageLoadStatus = evalPageLoad(m.pageLoadTime);
  m.ttiStatus      = evalTTI(m.timeToInteractive);

  const statuses = [m.lcpStatus, m.fcpStatus, m.clsStatus, m.ttfbStatus, m.pageLoadStatus, m.ttiStatus];
  const allGood = statuses.every(s => s === 'GOOD' || s === 'N/A');
  const anyPoor = statuses.some(s => s === 'POOR');
  m.overallStatus = anyPoor ? 'POOR' : allGood ? 'GOOD' : 'NEEDS IMPROVEMENT';
}

function evalLCP(v: number): BenchmarkStatus {
  if (v <= 0) return 'N/A';
  return v <= 2500 ? 'GOOD' : v <= 4000 ? 'NEEDS IMPROVEMENT' : 'POOR';
}
function evalFCP(v: number): BenchmarkStatus {
  if (v <= 0) return 'N/A';
  return v <= 1800 ? 'GOOD' : v <= 3000 ? 'NEEDS IMPROVEMENT' : 'POOR';
}
function evalCLS(v: number): BenchmarkStatus {
  if (v < 0) return 'N/A';
  return v <= 0.1 ? 'GOOD' : v <= 0.25 ? 'NEEDS IMPROVEMENT' : 'POOR';
}
function evalTTFB(v: number): BenchmarkStatus {
  if (v <= 0) return 'N/A';
  return v <= 800 ? 'GOOD' : v <= 1800 ? 'NEEDS IMPROVEMENT' : 'POOR';
}
function evalPageLoad(v: number): BenchmarkStatus {
  if (v <= 0) return 'N/A';
  return v <= 3000 ? 'GOOD' : v <= 6000 ? 'NEEDS IMPROVEMENT' : 'POOR';
}
function evalTTI(v: number): BenchmarkStatus {
  if (v <= 0) return 'N/A';
  return v <= 3800 ? 'GOOD' : v <= 7300 ? 'NEEDS IMPROVEMENT' : 'POOR';
}

// ── toString helper ────────────────────────────────────────────────────────────

export function metricsToString(m: PerformanceMetrics): string {
  return `[${m.pageName}] Load=${m.pageLoadTime.toFixed(0)}ms | FCP=${m.firstContentfulPaint.toFixed(0)}ms | LCP=${m.largestContentfulPaint.toFixed(0)}ms | TTI=${m.timeToInteractive.toFixed(0)}ms | CLS=${m.cumulativeLayoutShift.toFixed(3)} | TTFB=${m.timeToFirstByte.toFixed(0)}ms | AG-Grid=${m.agGridLoadTime.toFixed(0)}ms | Overall=${m.overallStatus}`;
}
